### ISSUE TITLE
create a wait_group_wrapper to avoid bug

### ISSUE THREAD

**Issue by hawkingrei** [pingcap/tidb#29517]
## Enhancement

In TiDB, we ignore the ```Done```  order so that it leads to some bug such as #29489.

We may create a wait_group_wrapper to automatically ```Done``` in the end. It is inspired by [nsq](https://github.com/nsqio/nsq/blob/master/internal/util/wait_group_wrapper.go#L7) and looks like this.

With its help, we can simplify the code.

```
        var wg sync.WaitGroup
        wg.Add(1)
	go func() {
		defer func() {
			close(writer.errCh)
			// close errCh before metaCh closed
			writer.wg.Done()
		}()
	        ......
	}()
```
Convert to
```
        var wg sync.WaitGroupWrapper
        wg.Wrap(func() {
               defer close(writer.errCh)
               ......
       })
```

**Comment by kennytm** [comment]
what about using [`errgroup`](https://pkg.go.dev/golang.org/x/sync/errgroup)?

**Comment by hawkingrei** [comment]
> what about using [`errgroup`](https://pkg.go.dev/golang.org/x/sync/errgroup)?

I think it can be used. but the ```wait_group_wrapper``` is more simple when it is without error return.

**Comment by hawkingrei** [comment]
```wait_group_wrapper``` has the added advantage of being compatible with ```WaitGroup```.

**Comment by YuJuncen** [comment]
Seems when the error is always `nil`, `errgroup` is isomorphic to `wait_group_wrapper`? 🤔

**Comment by YuJuncen** [comment]
ErrorGroup.Go:
```go
func (g *Group) Go(f func() error) {
	g.wg.Add(1)

	go func() {
		defer g.wg.Done()

		if err := f(); err != nil {
			g.errOnce.Do(func() {
				g.err = err
				if g.cancel != nil {
					g.cancel()
				}
			})
		}
	}()
}
```

WaitGroupWrapper.Wrap
https://github.com/nsqio/nsq/blob/198a8840f767dacb4ccbc0c560aadc25125e8e39/internal/util/wait_group_wrapper.go#L11-L17

**Comment by hawkingrei** [comment]
> Seems when the error is always `nil` and no `context` provided, `errgroup` is isomorphic to `wait_group_wrapper`? 🤔

Yes

errGroup
```
type Group struct {
	cancel func()

	wg sync.WaitGroup

	errOnce sync.Once
	err     error
}
```

WaitGroupWrapper
```
type WaitGroupWrapper struct {
	sync.WaitGroup
}
```

Comparison by struct definition, WaitGroupWrapper is more simple and compatible with ```WaitGroup```.

**Comment by kennytm** [comment]
Note that the nsq WaitGroupWrapper was introduced in nsqio/nsq#14 (2012-08-20), before `errgroup` was introduced on 2016-07-16 https://cs.opensource.google/go/x/sync/+/457c5828408160d6a47e17645169cf8fa20218c4.

**Comment by hawkingrei** [comment]
```errgroup``` provides synchronization, error propagation, and Context cancelation for groups of goroutines working on subtasks of a common task. If you need not context cancelation and so on, you should use ```WaitGroupWrapper``` to keep it simple.

**Comment by Zh1Cheung** [comment]
> Seems when the error is always `nil` and no `context` provided, `errgroup` is isomorphic to `wait_group_wrapper`? 🤔

If it was me, will consider encapsulating on errgroup instead of redesigning a structure similar to errgourp.

**Comment by hawkingrei** [comment]
> > Seems when the error is always `nil` and no `context` provided, `errgroup` is isomorphic to `wait_group_wrapper`? 🤔
> 
> Yes
> 
> errGroup
> 
> ```
> type Group struct {
> 	cancel func()
> 
> 	wg sync.WaitGroup
> 
> 	errOnce sync.Once
> 	err     error
> }
> ```
> 
> WaitGroupWrapper
> 
> ```
> type WaitGroupWrapper struct {
> 	sync.WaitGroup
> }
> ```
> 
> Comparison by struct definition, WaitGroupWrapper is more simple and compatible with `WaitGroup`.

```wait_group_wrapper``` is isomorphic to ```WaitGroup```.

**Comment by kennytm** [comment]
implemented in #29614 already, closing.
